documentation/modules/managing/con-producer-config-properties.adoc

```diff
@@ -0,0 +1,203 @@
+
+It’s likely that the messaging semantics (message ordering and durability) of your producer are defined by the requirements for your application.
+For instance, it’s possible that you don’t have the option of using `acks=0` or `acks=1` without breaking some important property or guarantee provided by your application.
+
+Broker restarts have a significant impact on high percentile statistics.
+For example, over a long period the 99th percentile latency is dominated by behavior around broker restarts.
+This is worth considering when designing benchmarks or comparing performance numbers from benchmarking with performance numbers seen in production.
+
+Depending on your objective, Kafka offers a number of configuration parameters and techniques for tuning producer performance for throughput and latency.
+
+Message batching (`linger.ms` and `batch.size`)::
+Message batching delays sending messages in the hope that more messages destined for the same broker will be sent,
+allowing them to be batched into a single produce request.
+Batching is a compromise between higher latency in return for higher throughput.
+Time-based batching is configured using `linger.ms`, and size-based batching is configured using `batch.size`.
+
+Compression (`compression.type`)::
+Message compression adds latency in the producer (CPU time spent compressing the messages),
+but makes requests (and potentially disk writes) smaller, which can increase throughput.
+Whether compression is worthwhile, and the best compression to use, will depend on the messages being sent.
+Compression happens on the thread which calls `KafkaProducer.send()`,
+so if the latency of this method matters for your application you should consider using more threads.
+
+Pipelining (`max.in.flight.requests.per.connection`)::
+Pipelining means sending more requests before the response to a previous request has been received.
+In general more pipelining means better throughput, up to a threshold at which other effects,
+such as worse batching, start to counteract the effect on throughput.
+
+.Lowering latency
+
+Producers scaled to produce a high rate of requests might cause queues in the Kafka cluster and increase latency.
+
+Use a buffer and add a small delay to reduce the number of requests to topics from producers that are not contributing a large number of messages.
+You can set the delay in conjunction with a maximum batch size, so the delay only applies if the batch size is not reached earlier.
+
+NOTE: If the system fails, there is a risk of unsent data in the buffer being lost.
+
+----
+# ...
+linger.ms=0 <1>
+batch.size=16384 <2>
+buffer.memory=33554432 <3>
+# ...
+----
+<1> The `linger` property adds a delay in milliseconds so that larger batches of messages are accumulated and sent in a request. The defaullt is `0'.`
+<2> If a maximum `batch.size` in bytes is used, a request is sent when the maximum is reached, or messages have been queued for longer than `linger.ms` (whichever comes sooner).
```

Comment:
in the above example, I don't think we are explaining well the balance between linger time and batch size. I mean, with above example, the message will be sent immediately because linger is 0, right?
If we want to explain better, maybe we should see linger to 100 ms for example and saying that messages are sent when 100 ms are reached without filling the batch size completely, or if you fill the entire batch size in less than 100 ms. @tombentley wdyt?

Reply:
sorry @tombentley AFAIK piggy-backing could mean delaying when talking about packet transmission not anticipating as you are describing. So what do you mean by piggy-backing in this case?

Reply:
We're sending some messages batches to a broker anyway, so we may as well add some partially full batch destined for the same broker. In a way it's not unrelated to the acknowledgement case you describe. In that case an ack is delayed so that we can ack two or more data frames with a single ack frame. TBH, I don't know how standardised my usage is here.

Reply:
Thanks for sharing this! ... but I see two possible different behaviors.
1. More partitions belong to the same broker. One batch for a partition is full and it's ready to be sent but ... the producer sends even the batches for other partitions belonging to the same broker even if they are not totally filled. In this case, we are anticipating acks.
2. More partitions belong to the same broker. One batch for a partition is full and it's ready to be sent but ... producer waits for the other batches belonging to the other partitions (but still on the same broker) to be filled before sending all of them. In this case, we are delaying acks.

Reply:
I changed the example to `linger.ms=100`

And we leave out the idea of piggy-backing here?

Reply:
I would leave it out, anyway it's not still clear how piggy-backing works in the producer.